oauth_tools/external_idp.py

```diff
@@ -147,23 +150,28 @@ def _apply_dex_resources(self) -> None:
             self._client.apply(obj, force=True)
 
         logger.info("Restarting dex")
-        self._restart_dex()
+        deleted_pod_names = self._restart_dex()
 
         logger.info("Waiting for dex to be ready")
-        self._wait_until_is_ready()
-
-    def __wait_until_is_ready(self) -> None:
-        for pod in self._client.list(Pod, namespace=self.namespace, labels={"app": "dex"}):
-            # Some pods may be deleted, if we are restarting
-            try:
+        self._wait_until_is_ready(ignore=deleted_pod_names)
+
+    def __wait_until_is_ready(self, ignore: Optional[List[str]] = None) -> None:
+        if ignore is None:
+            ignore = []
+        ready = False
+        while not ready:
+            for pod in self._client.list(Pod, namespace=self.namespace, labels={"app": "dex"}):
+                # Some pods may be deleted, if we are restarting
+                if pod.metadata.name in ignore:
+                    continue
                 self._client.wait(
                     Pod,
                     pod.metadata.name,
-                    for_conditions=["Ready", "Deleted"],
+                    for_conditions=["Ready"],
                     namespace=self.namespace,
                 )
-            except ObjectDeleted:
```

Comment:
(I might be totally wrong): So the `wait` API never throws the exception `ObjectDeleted`    for deleted Pods?

According to the following `wait` source code:

```python
        for op, obj in self.watch(res, namespace=namespace, fields={'metadata.name': name}):
            if obj.status is None:
                continue

            if op == "DELETED":
                raise ObjectDeleted(full_name)

            try:
                status = obj.status.to_dict()
            except AttributeError:
                status = obj.status

            conditions = [c for c in status.get('conditions', []) if c['status'] == 'True']
            if any(c['type'] in for_conditions for c in conditions):
                return obj

            failures = [c for c in conditions if c['type'] in raise_for_conditions]

            if failures:
                raise ConditionError(full_name, [f.get('message', f['type']) for f in failures])
```

Seems like it should raise the `ObjectDeleted` exception to signal the caller that the resource is no longer available.

Reply:
Or probably the `op == "DELETED` is only for the case when the pod is deleted after it has been waited?

Reply:
It will be true if the resources still exist in the Kubernetes cluster, and in that case the deleted event will be emitted.  The issue here is that when trying to invoke the `wait` function on a non-existent resource (already deleted). Under that condition, the `wait` function will hang indefinitely. In that case, it is equivalent to the code snippet below, and you can try this code to see that it will never return:

```python
import lightkube
from lightkube.resources.core_v1 import Pod

client = lightkube.Client()

client.wait(
    Pod,
    "non-existed-pod",
    for_conditions=["Ready", "Deleted"],
)
```

Since this issue is a race condition, it depends on the environment and cannot be 100% reproduced. In my environment, approximately 30% of test runs will encounter this problem.